kano_profile/badges.py

```diff
@@ -69,6 +69,29 @@ def calculate_xp():
     return int(points) + qm.evaluate_xp()
 
 
+def calculate_app_progress(app_name):
+    def get_group_progress(group_profile):
+        if 'challengeNo' not in group_profile:
+            return 0
+
+        return group_profile['challengeNo']
+
+    app_profile = load_app_state(app_name)
+    progress = 0
+
+    if 'groups' in app_profile and app_profile['groups']:
+        groups_profile = app_profile['groups']
+        for group in groups_profile:
+            progress += get_group_progress(groups_profile[group])
+
+        return progress
+
+    if 'level' in app_profile and app_profile['level']:
+        progress = app_profile['level']
+
+    return progress
+
+
 def calculate_kano_level():
```

Comment:
More of a pedantic comment here. Perhaps in the future we could model this into the user profile better than a function with so many helpers inside of it. Looks good for now though.